File: FrontEnd/ChromeExtension/jtoken-converter.js
import { deserialize, getPropertyName, indent as indentUtil } from './util.js';

let stringBuilder = '';

export default function objectInitializerUtil(jStr) {
    stringBuilder = '';
    let jObj = deserialize(jStr);
    constructObject(jObj, 0, true);
    return stringBuilder;
}

function constructObject(jObj, level, isLast = false, isNewLine = true) {
    let objectType = typeof jObj;
    if (objectType == 'object' && jObj != null && !jObj.length) {
        indent(level, isNewLine);
        stringBuilder += 'new JObject\n';
        indent(level);
        stringBuilder += '{\n';
        let keys = Object.getOwnPropertyNames(jObj);
        keys.forEach((key, index, _keys) => {
            let child = jObj[key];

            let childType = typeof child;
            indent(level + 1, true);
            stringBuilder += 'new JProperty(';
            stringBuilder += '"' + getPropertyName(key) + '"';
            stringBuilder += ', ';

            let isLastChild = keys.length - 1 == index;
            constructObject(child, level + 1, isLastChild, false);
            stringBuilder += ')';

            if (!isLastChild && (child == null || childType != 'object')) {
                stringBuilder += ',\n';
            }
        });

        stringBuilder += '\n';
        indent(level);
        stringBuilder += '}';
        if (!isLast) {
            stringBuilder += ',\n';
        }
    } else if (objectType == 'object' && jObj != null && jObj.length) {
        indent(level, isNewLine);
        stringBuilder += 'new JArray\n';
        indent(level);
        stringBuilder += '{\n';
        jObj.forEach((child, index, _jObj) => {
            let childType = typeof child;
            let isLastChild = jObj.length - 1 == index;
            constructObject(child, level + 1, isLastChild);
            if (!isLastChild && (childType != 'object')) {
                stringBuilder += ',\n';
            }
        });

        stringBuilder += '\n';
        indent(level);
        stringBuilder += '}';
        if (!isLast) {
            stringBuilder += ',\n';
        }
    } else if (objectType == 'string') {
        indent(level, isNewLine);
        var stringLiteral = `"${jObj}"`;
        stringBuilder += stringLiteral;
    } else if (objectType == 'number' || objectType == 'boolean' || jObj == null) {
        indent(level, isNewLine);
        stringBuilder += jObj;
    } else {
        throw "Invalid element type.";
    }
}

function indent(tabIndentCount, isNewLine = true) {
    stringBuilder += indentUtil(tabIndentCount, isNewLine);
}

File: BackEnd/JSON-to-ObjectInitializer/JTokenConverter.cs
namespace JSON_to_ObjectInitializer
{
    using Newtonsoft.Json.Linq;
    using System.Text;

    public class JTokenConverter : IConverter
    {
        private StringBuilder Sb { get; set; } = new StringBuilder();

        public string ConstructObjectInitializerFormat(JToken jObj)
        {
            // Construction based on JSON.NET types
            this.ConstructObject(jObj, 0, isLast: true);
            string result = this.Sb.ToString();
            result = result.Replace(@"bool~True", "true");
            result = result.Replace(@"bool~False", "false");
            return result;
        }

        private void ConstructObject(JToken jObj, int level, bool isLast = false, bool isNewLine = true)
        {
            if (jObj.Type == JTokenType.Object)
            {
                this.Indent(level, isNewLine);
                this.Sb.Append("new JObject\n");
                this.Indent(level);
                this.Sb.Append("{\n");
                foreach (JToken child in jObj.Children())
                {
                    bool isLastChild = child.Next == null;
                    this.ConstructObject(child, level + 1, isLastChild);
                }

                this.Sb.Append("\n");
                this.Indent(level);
                this.Sb.Append("}");
                if (!isLast)
                {
                    this.Sb.Append(",\n");
                }
            }
            else if (jObj.Type == JTokenType.Array)
            {
                this.Indent(level, isNewLine);
                this.Sb.Append("new JArray\n");
                this.Indent(level);
                this.Sb.Append("{\n");
                foreach (JToken child in jObj.Children())
                {
                    bool isLastChild = child.Next == null;
                    this.ConstructObject(child, level + 1, isLastChild);
                    if (child.Next != null && (child.Type != JTokenType.Object && child.Type != JTokenType.Array && child.Type != JTokenType.Property))
                    {
                        this.Sb.Append(",\n");
                    }
                }

                this.Sb.Append("\n");
                this.Indent(level);
                this.Sb.Append("}");
                if (!isLast)
                {
                    this.Sb.Append(",\n");
                }
            }
            else if (jObj.Type == JTokenType.Property)
            {
                this.Indent(level);
                this.Sb.Append("new JProperty(");
                foreach (JToken child in jObj.Children())
                {
                    this.Sb.Append(string.Format("\"{0}\"", JsonSerializerUtil.GetPropertyName(child.Path)));
                    this.Sb.Append(", ");
                    this.ConstructObject(child, level, true, false);
                }

                this.Sb.Append(")");
                if (!isLast)
                {
                    this.Sb.Append(",\n");
                }
            }
            else if (jObj.Type == JTokenType.String)
            {
                this.Indent(level, isNewLine);
                var stringLiteral = string.Format("\"{0}\"", jObj.Value<string>());
                this.Sb.Append(stringLiteral);
            }
            else if (jObj.Type == JTokenType.Integer)
            {
                this.Indent(level, isNewLine);
                this.Sb.Append(jObj.Value<int>());
            }
            else if (jObj.Type == JTokenType.Float)
            {
                this.Indent(level, isNewLine);
                this.Sb.Append(jObj.Value<float>());
            }
            else if (jObj.Type == JTokenType.Boolean)
            {
                this.Indent(level, isNewLine);

                // Adding `bool~` to the value to have proper boolean literal in the output
                // Ex: To change False/True to false/true
                this.Sb.Append("bool~" + jObj.Value<bool>());
            }
            else if (jObj.Type == JTokenType.Null)
            {
                this.Indent(level, isNewLine);
                var nullLiteral = "null";
                this.Sb.Append(nullLiteral);
            }
        }

        private void Indent(int tabIndentCount, bool isNewLine = true)
        {
            this.Sb.Append(JsonSerializerUtil.Indent(tabIndentCount, isNewLine));
        }
    }
}
